### PR TITLE
Update Migration-Guide-V4.md

### LINKED ISSUE #4085: Hook onRoute doesn't work correct inside registered plugin
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.1.0

### Plugin version

_No response_

### Node.js version

16.15.1

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

Ubuntu 22.04 LTS

### Description

If the onRoute hook is in the main application code, then callback is called for all routes. And extraneous HEAD routes appear. If the onRoute hook is placed inside the plugin, then the callback is called only for the HEAD routes. But registration of such a HEAD route did not occur.

### Steps to Reproduce

Example:

```javascript
const fastify = Fastify({ logger: true });

fastify.addHook('onRoute', (routeOptions) => {
  const { path, method } = routeOptions;
  console.log({ path, method });
});

fastify.get('/', (req, reply) => {
  reply.send('index');
});

// Output:
// { path: '/', method: 'GET' }
// { path: '/', method: 'HEAD' }
```

Example with hook inside plugin:

```javascript
const fastify = Fastify({ logger: true });

fastify.register((instance, opts, done) => {
  instance.addHook('onRoute', (routeOptions) => {
    const { path, method } = routeOptions;
    console.log({ path, method });
  });
  done();
});

fastify.get('/', (req, reply) => {
  reply.send('index');
});

// Output:
// Empty output
```

Example with using fastify-plugin

```javascript
import fp from 'fastify-plugin';

const fastify = Fastify({ logger: true });

fastify.register(fp((instance, opts, done) => {
  instance.addHook('onRoute', (routeOptions) => {
    const { path, method } = routeOptions;
    console.log({ path, method });
  });
  done();
}));

fastify.get('/', (req, reply) => {
  reply.send('index');
});

// Output:
// { path: '/', method: 'HEAD' }
// No GET route
```

### Expected Behavior

Inside the plugin, the hook should have a route `{ path: '/', method: 'GET' }` and should not have a HEAD route. When using a hook outside the plugin, there should also be only one GET route.
This behavior was in fastify version 3.29.0.

### PR DESCRIPTION
**Pull request by denes-fekeshazy**
Extending the V4 migration guide to highlight the changes to route definitions.

Resolves #4085 

#### Checklist

- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Guides/Migration-Guide-V4.md [hunk 1]
```diff
@@ -65,6 +65,14 @@ Starting from v4, all the `GET` routes will create a sibling `HEAD` route.
 You can revert this behaviour by setting the server's option `exposeHeadRoutes`
 to `false`.
 
+### Synchronous route definitions
+
+The route registration has been made syncrounes from v4. This change was done to provide
```
Comment (mcollina) [inline docs/Guides/Migration-Guide-V4.md:70]:
```suggestion
The route registration has been made synchronous from v4. This change was done to provide
```

File: docs/Guides/Migration-Guide-V4.md [hunk 2]
```diff
@@ -65,6 +65,14 @@ Starting from v4, all the `GET` routes will create a sibling `HEAD` route.
...
+better error reporting for route definition. As a result if you specify an `onRoute` hook
+in a plugin you should either:
+* wrap your routes in a plugin (recommended)
+* use `await register(...)`
```
Comment (Eomm) [inline docs/Guides/Migration-Guide-V4.md:74]: I would add the code example to let the user understand better the implications:
any `onRoute` hooks will not be executed when a plugin adds them if it is not awaited or  the routes are wrapped in a plugin

Comment (denes-fekeshazy) [reply docs/Guides/Migration-Guide-V4.md:74]: I added the code example used in the linked issue before and after refactoring. Could you have a look please if that was what you had in mind?

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**